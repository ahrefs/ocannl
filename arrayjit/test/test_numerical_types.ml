open Base
module Ndarray = Ir.Ndarray
module Ops = Ir.Ops

let test_bfloat16_conversions () =
  Stdio.printf "Testing BFloat16 conversions:\n";

  (* Test some specific values *)
  let test_values = [ 0.0; 1.0; -1.0; 3.14159; 1e-3; 1e3; Float.infinity; Float.neg_infinity ] in

  List.iter test_values ~f:(fun orig ->
      let bf16 = Ndarray.float_to_bfloat16 orig in
      let back = Ndarray.bfloat16_to_float bf16 in
      Stdio.printf "  %.6f -> 0x%04x -> %.6f\n" orig bf16 back);

  (* Test round-trip through ndarray *)
  let arr =
    Ndarray.create_array ~debug:"test" Ops.bfloat16 ~dims:[| 3; 2 |]
      (Ops.Constant_fill { values = [| 1.0; 2.0; 3.14; -1.5; 0.125; 1000.0 |]; strict = true })
  in

  Stdio.printf "\nBFloat16 array values:\n";
  let flat_values = Ndarray.retrieve_flat_values arr in
  Array.iteri flat_values ~f:(fun i v -> Stdio.printf "  [%d] = %.6f\n" i v)

let test_fp8_conversions () =
  Stdio.printf "\n\nTesting FP8 conversions:\n";

  (* Test some specific values *)
  let test_values = [ 0.0; 1.0; -1.0; 0.5; 2.0; 0.125; 16.0; -0.25 ] in

  List.iter test_values ~f:(fun orig ->
      let fp8 = Ndarray.float_to_fp8 orig in
      let back = Ndarray.fp8_to_float fp8 in
      Stdio.printf "  %.6f -> 0x%02x -> %.6f\n" orig fp8 back);

  (* Test round-trip through ndarray *)
  let arr =
    Ndarray.create_array ~debug:"test" Ops.fp8 ~dims:[| 2; 2 |]
      (Ops.Constant_fill { values = [| 1.0; 0.5; 2.0; -1.0 |]; strict = true })
  in

  Stdio.printf "\nFP8 array values:\n";
  let flat_values = Ndarray.retrieve_flat_values arr in
  Array.iteri flat_values ~f:(fun i v -> Stdio.printf "  [%d] = %.6f\n" i v)

let () =
  test_bfloat16_conversions ();
  test_fp8_conversions ()
