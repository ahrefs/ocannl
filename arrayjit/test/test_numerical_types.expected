
Welcome to OCANNL! Reading configuration defaults from /Users/lukstafi/ocannl/_build/default/arrayjit/test/ocannl_config.
Retrieving commandline, environment, or config file variable ocannl_log_level
Found 0, in the config file
Testing BFloat16 conversions:
  0.000000 -> 0x0000 -> 0.000000
  1.000000 -> 0x3f80 -> 1.000000
  -1.000000 -> 0xbf80 -> -1.000000
  3.141590 -> 0x4049 -> 3.140625
  0.001000 -> 0x3a83 -> 0.000999
  1000.000000 -> 0x447a -> 1000.000000
  inf -> 0x7f80 -> inf
  -inf -> 0xff80 -> -inf

BFloat16 array values:
  [0] = 1.000000
  [1] = 2.000000
  [2] = 3.140625
  [3] = -1.500000
  [4] = 0.125000
  [5] = 1000.000000


Testing FP8 conversions:
  0.000000 -> 0x00 -> 0.000000
  1.000000 -> 0x3c -> 1.000000
  -1.000000 -> 0xbc -> -1.000000
  0.500000 -> 0x38 -> 0.500000
  2.000000 -> 0x40 -> 2.000000
  0.125000 -> 0x30 -> 0.125000
  16.000000 -> 0x4c -> 16.000000
  -0.250000 -> 0xb4 -> -0.250000

FP8 array values:
  [0] = 1.000000
  [1] = 0.500000
  [2] = 2.000000
  [3] = -1.000000
